plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/loops/SelfHyperLoopLabels.java

```diff
@@ -0,0 +1,257 @@
+        LEFT,
+        /** A northern or southern right-aligned label. */
+        RIGHT,
+        /** An eastern or western top-aligned label. */
+        TOP
+    }
+
+    // CHECKSTYLEOFF VisibilityModifier
+    /** An ID that can be used arbitrarily. */
+    public int id = 0;
+    // CHECKSTYLEON VisibilityModifier
+    
+    /** The labels represented by this instance. */
+    private final List<LLabel> lLabels = new ArrayList<>();
+    /** The size required to place all labels. */
+    private final KVector size = new KVector();
+    /** The position of our bounding box's top left corner. */
+    private final KVector position = new KVector();
+    /** The graph's layout direction. Influences the way the labels will be placed. */
+    private final Direction layoutDirection;
+    /** Space to leave between adjacent labels. */
+    private final double labelLabelSpacing;
+    
+    /** The side the label is placed on. */
+    private PortSide side;
+    /** The label's alignment. */
+    private Alignment alignment;
+    /** The port any non-center alignment is relative to. */
+    private SelfLoopPort alignmentReferenceSLPort;
+    
+    
+    /**
+     * Creates a new instance for the given {@link SelfHyperLoop}.
+     */
+    public SelfHyperLoopLabels(final SelfHyperLoop slLoop) {
+        // Initialize our properties from the graph
+        LNode lNode = slLoop.getSLHolder().getLNode();
+        
+        layoutDirection = lNode.getGraph().getProperty(LayeredOptions.DIRECTION);
+        labelLabelSpacing = LGraphUtil.getIndividualOrInherited(lNode, LayeredOptions.SPACING_LABEL_LABEL);
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // LLabel Access
+    
+    /**
+     * Adds the given collection of labels to this instance.
+     */
+    public void addLLabels(final Collection<LLabel> newLLabels) {
+        for (LLabel newLLabel : newLLabels) {
+            lLabels.add(newLLabel);
+            updateSize(newLLabel);
+        }
+    }
+
+    /**
+     * Returns the labels represented by this instance.
+     */
+    public List<LLabel> getLLabels() {
+        return lLabels;
+    }
+    
+    /**
+     * Updates our size under the assumption that the given label was just added to this instance.
+     */
+    private void updateSize(final LLabel newLLabel) {
+        assert lLabels.get(lLabels.size() - 1) == newLLabel;
+        
+        KVector newLLabelSize = newLLabel.getSize();
+        
+        if (layoutDirection.isHorizontal()) {
+            // The labels will be stacked vertically
+            size.x = Math.max(size.x, newLLabelSize.x);
+            size.y += newLLabelSize.y;
+            
+            // Add a label-label spacing if we already had labels
+            if (lLabels.size() > 1) {
+                size.y += labelLabelSpacing;
+            }
+            
+        } else {
+            // The labels will be stacked horizontally
```

Comment:
Does this make sense as the common case? The text is still horizontal, right? Hence the drawing may become quite wide?

Reply:
Nope, this code is only executed for vertical layouts. Due to the `GraphTransformer`, the labels will then be narrow, but high, with their text (conceptually) running vertically.